src/kernels/raw/finder/localKernelFinder.node.ts

```diff
@@ -4,134 +4,226 @@
-            resource,
-            cancelToken
-        );
+    private async loadCache() {
+        // loading cache, which is resource agnostic
+        const kernelsFromCache = await this.getFromCache();
         let kernels: LocalKernelConnectionMetadata[] = [];
-        if (useCache === 'ignoreCache') {
+
+        if (Array.isArray(kernelsFromCache) && kernelsFromCache.length > 0) {
+            kernels = kernelsFromCache;
+        } else {
             try {
-                kernels = await kernelsWithoutCachePromise;
+                kernels = await this.listKernels(undefined);
             } catch (ex) {
-                traceWarning(`Could not fetch kernels from the ${this.kind}`);
-                kernels = [];
-                updateCache = false;
+                traceError(`Exception loading kernels: ${ex}`);
             }
-        } else {
-            let kernelsFromCache: LocalKernelConnectionMetadata[] | undefined;
-            kernelsFromCachePromise
-                .then((items) => {
-                    kernelsFromCache = items;
-                    updateCache = false;
-                })
-                .catch(noop);
+        }
+        await this.writeToCache(kernels);
+        this._initializeResolve();
+    }
+
+    private async updateCache(resource: Resource) {
+        this._cacheUpdateCancelTokenSource?.dispose();
+        const updateCacheCancellationToken = new CancellationTokenSource();
+        this._cacheUpdateCancelTokenSource = updateCacheCancellationToken;
+
+        if (resource) {
+            let kernels: LocalKernelConnectionMetadata[];
 
             try {
-                const kernelsWithoutCacheDeferred = createDeferredFromPromise(kernelsWithoutCachePromise);
-                try {
-                    await Promise.race([kernelsFromCachePromise, kernelsWithoutCacheDeferred.promise]);
-                } catch (ex) {
-                    // If we failed to get without cache, then await on the cache promise as a fallback.
-                    if (kernelsWithoutCacheDeferred.rejected) {
-                        await kernelsFromCachePromise;
-                    } else {
-                        throw ex;
-                    }
-                }
-                // If we finish the cache first, and we don't have any items, in the cache, then load without cache.
-                if (Array.isArray(kernelsFromCache) && kernelsFromCache.length > 0) {
-                    kernels = kernelsFromCache;
-                } else {
-                    kernels = await kernelsWithoutCachePromise;
-                    updateCache = true;
-                }
+                kernels = await this.listKernels(resource, updateCacheCancellationToken.token);
             } catch (ex) {
                 traceError(`Exception loading kernels: ${ex}`);
+                return;
             }
+
+            // update resource cache
+            const resourceCacheKey = this.getResourceCacheKey(resource);
```

Comment:
You'd need to return from here if the cancellation token has been cancelled.
Else we end up updating emptu items into the list and triggering the event.